Incorrectly marshaled LPCWSTR

##### Brief Description
I think I found a bug. Type LPCWSTR is defined as:
> An LPCWSTR is a 32-bit pointer to a constant string of 16-bit Unicode characters, which MAY be null-terminated. This type is declared as follows:
> typedef const wchar_t* LPCWSTR;

parameter of this type generates the following code:
`[MarshalAs(UnmanagedType.LPUTF8Str)] string paramName`

I think it should have generated:
`[MarshalAs(UnmanagedType.LPWStr)] string paramName`

OS: Microsoft Windows 10 64-bit


##### Used headers
https://github.com/microsoft/webauthn/blob/master/webauthn.h


##### Used settings

Target: MSVC

```
driver.ParserOptions.ForceClangToolchainLookup = true;
driver.ParserOptions.Verbose = true;
options.GeneratorKind = GeneratorKind.CSharp;
options.Verbose = true;
```

@ddobrev I'd be happy to do that, but I can't get the source to build locally. I took steps according to [tutorial](https://github.com/mono/CppSharp/blob/master/docs/GettingStarted.md), tried it twice, but the build still doesn't go through. Here's the output:

```
1>------ Build started: Project: CppSharp.CppParser, Configuration: Debug x64 ------
2>------ Build started: Project: CppSharp.Runtime, Configuration: Debug x64 ------
3>------ Build started: Project: CppSharp, Configuration: Debug x64 ------
4>------ Build started: Project: CppSharp.AST, Configuration: Debug x64 ------
5>------ Build started: Project: Std-symbols, Configuration: Debug x64 ------
6>------ Build started: Project: NamespacesBase.Native, Configuration: Debug x64 ------
7>------ Build started: Project: CSharp.Native, Configuration: Debug x64 ------
8>------ Build started: Project: VTables.Native, Configuration: Debug x64 ------
2>  CppSharp.Runtime -> F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Runtime.dll
3>  CppSharp -> F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.dll
9>------ Build started: Project: CLI.Native, Configuration: Debug x64 ------
4>  CppSharp.AST -> F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.AST.dll
10>------ Build started: Project: StandardLib.Native, Configuration: Debug x64 ------
11>------ Build started: Project: Common.Native, Configuration: Debug x64 ------
7>AnotherUnit.cpp
6>NamespacesBase.cpp
10>StandardLib.cpp
11>AnotherUnit.cpp
9>CLI.cpp
5>Std-symbols.cpp
11>Common.cpp
1>AST.cpp
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(24): warning C4251: 'TestVectors::IntVector': class 'std::vector<int,std::allocator<_Ty>>' needs to have dll-interface to be used by clients of struct 'TestVectors'
10>        with
10>        [
10>            _Ty=int
10>        ]
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(20): note: see declaration of 'std::vector<int,std::allocator<_Ty>>'
10>        with
10>        [
10>            _Ty=int
10>        ]
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(26): warning C4251: 'TestVectors::IntPtrVector': class 'std::vector<int *,std::allocator<_Ty>>' needs to have dll-interface to be used by clients of struct 'TestVectors'
10>        with
10>        [
10>            _Ty=int *
10>        ]
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(26): note: see declaration of 'std::vector<int *,std::allocator<_Ty>>'
10>        with
10>        [
10>            _Ty=int *
10>        ]
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(28): warning C4251: 'TestVectors::IntWrapperVector': class 'std::vector<IntWrapper,std::allocator<_Ty>>' needs to have dll-interface to be used by clients of struct 'TestVectors'
10>        with
10>        [
10>            _Ty=IntWrapper
10>        ]
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(28): note: see declaration of 'std::vector<IntWrapper,std::allocator<_Ty>>'
10>        with
10>        [
10>            _Ty=IntWrapper
10>        ]
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(30): warning C4251: 'TestVectors::IntWrapperPtrVector': class 'std::vector<IntWrapper *,std::allocator<_Ty>>' needs to have dll-interface to be used by clients of struct 'TestVectors'
10>        with
10>        [
10>            _Ty=IntWrapper *
10>        ]
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(30): note: see declaration of 'std::vector<IntWrapper *,std::allocator<_Ty>>'
10>        with
10>        [
10>            _Ty=IntWrapper *
10>        ]
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(32): warning C4251: 'TestVectors::IntWrapperValueTypeVector': class 'std::vector<IntWrapperValueType,std::allocator<_Ty>>' needs to have dll-interface to be used by clients of struct 'TestVectors'
10>        with
10>        [
10>            _Ty=IntWrapperValueType
10>        ]
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(15): note: see declaration of 'std::vector<IntWrapperValueType,std::allocator<_Ty>>'
10>        with
10>        [
10>            _Ty=IntWrapperValueType
10>        ]
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(34): warning C4251: 'TestVectors::IntWrapperValueTypeVectorTypedef': class 'std::vector<IntWrapperValueType,std::allocator<_Ty>>' needs to have dll-interface to be used by clients of struct 'TestVectors'
10>        with
10>        [
10>            _Ty=IntWrapperValueType
10>        ]
10>f:\temp\cppsharp\tests\standardlib\standardlib.h(15): note: see declaration of 'std::vector<IntWrapperValueType,std::allocator<_Ty>>'
10>        with
10>        [
10>            _Ty=IntWrapperValueType
10>        ]
8>VTables.cpp
9>f:\temp\cppsharp\tests\cli\cli.h(73): warning C4251: 'TestMappedTypeNonConstRefParam::m_str': class 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>' needs to have dll-interface to be used by clients of class 'TestMappedTypeNonConstRefParam'
9>c:\program files (x86)\microsoft visual studio\2017\enterprise\vc\tools\msvc\14.16.27023\include\xstring(4373): note: see declaration of 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>'
7>CSharp.cpp
11>f:\temp\cppsharp\tests\common\common.h(778): warning C4200: nonstandard extension used: zero-sized array in struct/union
11>f:\temp\cppsharp\tests\common\common.h(778): note: This member will be ignored by a defaulted constructor or copy/move assignment operator
11>f:\temp\cppsharp\tests\common\common.h(919): warning C4251: 'HasStdString::s': class 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>' needs to have dll-interface to be used by clients of class 'HasStdString'
11>c:\program files (x86)\microsoft visual studio\2017\enterprise\vc\tools\msvc\14.16.27023\include\xstring(4373): note: see declaration of 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>'
9>ClassWithNestedEnum.cpp
9>NestedEnumConsumer.cpp
9>Employee.cpp
8>f:\temp\cppsharp\tests\vtables\vtables.h(20): warning C4251: 'Foo::s': class 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>' needs to have dll-interface to be used by clients of class 'Foo'
8>c:\program files (x86)\microsoft visual studio\2017\enterprise\vc\tools\msvc\14.16.27023\include\xstring(4373): note: see declaration of 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>'
11>Generating Code...
11>f:\temp\cppsharp\tests\common\common.cpp(428): warning C4717: 'operator|': recursive on all control paths, function will cause runtime stack overflow
7>f:\temp\cppsharp\tests\csharp\csharp.h(52): warning C4251: 'Foo::_name': class 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>' needs to have dll-interface to be used by clients of class 'Foo'
7>c:\program files (x86)\microsoft visual studio\2017\enterprise\vc\tools\msvc\14.16.27023\include\xstring(4373): note: see declaration of 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>'
6>   Creating library ..\lib\Debug_x64\NamespacesBase.Native.lib and object ..\lib\Debug_x64\NamespacesBase.Native.exp
5>   Creating library ..\lib\Debug_x64\Std-symbols.lib and object ..\lib\Debug_x64\Std-symbols.exp
1>f:\temp\cppsharp\src\cppparser\ast.cpp(11): fatal error C1083: Cannot open include file: 'llvm/ADT/SmallString.h': No such file or directory
1>Comments.cpp
1>f:\temp\cppsharp\src\cppparser\parser.h(10): fatal error C1083: Cannot open include file: 'llvm/Object/Archive.h': No such file or directory
1>CppParser.cpp
7>f:\temp\cppsharp\tests\csharp\csharptemplates.h(122): warning C4244: 'return': conversion from 'float' to 'int', possible loss of data
7>f:\temp\cppsharp\tests\csharp\csharptemplates.h(121): note: while compiling class template member function 'int IndependentFields<T>::getIndependent(void)'
7>        with
7>        [
7>            T=int
7>        ]
7>f:\temp\cppsharp\tests\csharp\csharptemplates.h(543): note: see reference to class template instantiation 'IndependentFields<T>' being compiled
7>        with
7>        [
7>            T=int
7>        ]
11>   Creating library ..\lib\Debug_x64\Common.Native.lib and object ..\lib\Debug_x64\Common.Native.exp
7>CSharpTemplates.cpp
8>   Creating library ..\lib\Debug_x64\VTables.Native.lib and object ..\lib\Debug_x64\VTables.Native.exp
9>EmployeeOrg.cpp
5>Std-symbols.vcxproj -> F:\temp\CppSharp\build\vs2017\projects\..\lib\Debug_x64\Std-symbols.dll
6>NamespacesBase.Native.vcxproj -> F:\temp\CppSharp\build\vs2017\projects\..\lib\Debug_x64\NamespacesBase.Native.dll
12>------ Build started: Project: Encodings.Native, Configuration: Debug x64 ------
13>------ Build started: Project: NamespacesDerived.Native, Configuration: Debug x64 ------
11>Common.Native.vcxproj -> F:\temp\CppSharp\build\vs2017\projects\..\lib\Debug_x64\Common.Native.dll
8>VTables.Native.vcxproj -> F:\temp\CppSharp\build\vs2017\projects\..\lib\Debug_x64\VTables.Native.dll
11>Done building project "Common.Native.vcxproj".
8>Done building project "VTables.Native.vcxproj".
10>   Creating library ..\lib\Debug_x64\StandardLib.Native.lib and object ..\lib\Debug_x64\StandardLib.Native.exp
12>Encodings.cpp
1>f:\temp\cppsharp\src\cppparser\parser.h(10): fatal error C1083: Cannot open include file: 'llvm/Object/Archive.h': No such file or directory
9>Generating Code...
7>f:\temp\cppsharp\tests\csharp\csharptemplates.h(122): warning C4244: 'return': conversion from 'float' to 'int', possible loss of data
7>f:\temp\cppsharp\tests\csharp\csharptemplates.h(121): note: while compiling class template member function 'int IndependentFields<T>::getIndependent(void)'
7>        with
7>        [
7>            T=int
7>        ]
1>Expr.cpp
7>f:\temp\cppsharp\tests\csharp\csharptemplates.h(543): note: see reference to class template instantiation 'IndependentFields<T>' being compiled
7>        with
7>        [
7>            T=int
7>        ]
7>Generating Code...
10>StandardLib.Native.vcxproj -> F:\temp\CppSharp\build\vs2017\projects\..\lib\Debug_x64\StandardLib.Native.dll
10>Done building project "StandardLib.Native.vcxproj".
13>NamespacesDerived.cpp
12>f:\temp\cppsharp\tests\encodings\encodings.h(11): warning C4251: 'Foo::StringVariable': class 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>' needs to have dll-interface to be used by clients of class 'Foo'
12>c:\program files (x86)\microsoft visual studio\2017\enterprise\vc\tools\msvc\14.16.27023\include\xstring(4373): note: see declaration of 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>'
9>   Creating library ..\lib\Debug_x64\CLI.Native.lib and object ..\lib\Debug_x64\CLI.Native.exp
13>   Creating library ..\lib\Debug_x64\NamespacesDerived.Native.lib and object ..\lib\Debug_x64\NamespacesDerived.Native.exp
7>   Creating library ..\lib\Debug_x64\CSharp.Native.lib and object ..\lib\Debug_x64\CSharp.Native.exp
12>   Creating library ..\lib\Debug_x64\Encodings.Native.lib and object ..\lib\Debug_x64\Encodings.Native.exp
9>CLI.Native.vcxproj -> F:\temp\CppSharp\build\vs2017\projects\..\lib\Debug_x64\CLI.Native.dll
9>Done building project "CLI.Native.vcxproj".
13>NamespacesDerived.Native.vcxproj -> F:\temp\CppSharp\build\vs2017\projects\..\lib\Debug_x64\NamespacesDerived.Native.dll
1>ParseExpr.cpp
12>Encodings.Native.vcxproj -> F:\temp\CppSharp\build\vs2017\projects\..\lib\Debug_x64\Encodings.Native.dll
12>Done building project "Encodings.Native.vcxproj".
7>CSharp.Native.vcxproj -> F:\temp\CppSharp\build\vs2017\projects\..\lib\Debug_x64\CSharp.Native.dll
7>Done building project "CSharp.Native.vcxproj".
1>f:\temp\cppsharp\src\cppparser\parser.h(10): fatal error C1083: Cannot open include file: 'llvm/Object/Archive.h': No such file or directory
1>ParseStmt.cpp
1>f:\temp\cppsharp\src\cppparser\parser.h(10): fatal error C1083: Cannot open include file: 'llvm/Object/Archive.h': No such file or directory
1>Parser.cpp
1>f:\temp\cppsharp\src\cppparser\parser.h(10): fatal error C1083: Cannot open include file: 'llvm/Object/Archive.h': No such file or directory
1>Sources.cpp
1>Stmt.cpp
1>Target.cpp
1>Generating Code...
1>Done building project "CppSharp.CppParser.vcxproj" -- FAILED.
14>------ Build started: Project: CppSharp.Parser.CLI, Configuration: Debug x64 ------
15>------ Build started: Project: CppSharp.Parser.CSharp, Configuration: Debug x64 ------
14>AST.cpp
15>  CppSharp.Parser.CSharp -> F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CSharp.dll
14>CppParser.cpp
14>Decl.cpp
14>Expr.cpp
14>Sources.cpp
14>Stmt.cpp
14>Target.cpp
14>Types.cpp
14>Generating Code...
14>.NETFramework,Version=v4.5.2.AssemblyAttributes.cpp
14>LINK : fatal error LNK1104: cannot open file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.CppParser.lib'
14>Done building project "CppSharp.Parser.CLI.vcxproj" -- FAILED.
16>------ Build started: Project: CppSharp.Parser, Configuration: Debug x64 ------
16>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
17>------ Build started: Project: CppSharp.Generator, Configuration: Debug x64 ------
17>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
17>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
18>------ Build started: Project: CppSharp.Generator.Tests, Configuration: Debug x64 ------
19>------ Build started: Project: CppSharp.CLI, Configuration: Debug x64 ------
20>------ Build started: Project: CppSharp.Parser.Bootstrap, Configuration: Debug x64 ------
21>------ Build started: Project: CppSharp.Parser.Gen, Configuration: Debug x64 ------
21>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.dll' could not be found
21>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
21>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
19>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.dll' could not be found
19>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
19>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
18>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.dll' could not be found
18>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
18>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
22>------ Build started: Project: NamespacesDerived.Gen, Configuration: Debug x64 ------
23>------ Build started: Project: CLI.Gen, Configuration: Debug x64 ------
24>------ Build started: Project: Common.Gen, Configuration: Debug x64 ------
25>------ Build started: Project: CSharp.Gen, Configuration: Debug x64 ------
26>------ Build started: Project: Encodings.Gen, Configuration: Debug x64 ------
27>------ Build started: Project: StandardLib.Gen, Configuration: Debug x64 ------
28>------ Build started: Project: VTables.Gen, Configuration: Debug x64 ------
25>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.dll' could not be found
25>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
25>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
25>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
29>------ Build started: Project: CSharp.CSharp, Configuration: Debug x64 ------
24>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.dll' could not be found
24>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
24>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
24>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
30>------ Build started: Project: Common.CLI, Configuration: Debug x64 ------
23>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.dll' could not be found
23>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
23>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
23>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
26>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.dll' could not be found
26>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
26>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
26>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
31>------ Build started: Project: CLI.CLI, Configuration: Debug x64 ------
32>------ Build started: Project: Encodings.CSharp, Configuration: Debug x64 ------
28>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.dll' could not be found
28>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
28>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
28>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
33>------ Build started: Project: Common.CSharp, Configuration: Debug x64 ------
22>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.dll' could not be found
22>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
22>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
22>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
29>  '"F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/CSharp.Gen.exe"' is not recognized as an internal or external command,
29>  operable program or batch file.
29>C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(1330,5): error MSB3073: The command ""F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/CSharp.Gen.exe"" exited with code 9009.
34>------ Build started: Project: NamespacesBase.CSharp, Configuration: Debug x64 ------
32>  '"F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/Encodings.Gen.exe"' is not recognized as an internal or external command,
35>------ Build started: Project: VTables.CSharp, Configuration: Debug x64 ------
32>  operable program or batch file.
32>C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(1330,5): error MSB3073: The command ""F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/Encodings.Gen.exe"" exited with code 9009.
33>  '"F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/Common.Gen.exe"' is not recognized as an internal or external command,
33>  operable program or batch file.
33>C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(1330,5): error MSB3073: The command ""F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/Common.Gen.exe"" exited with code 9009.
36>------ Build started: Project: CSharp.Tests.CSharp, Configuration: Debug x64 ------
20>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.dll' could not be found
20>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
20>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
37>------ Build started: Project: Encodings.Tests.CSharp, Configuration: Debug x64 ------
27>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.dll' could not be found
27>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
27>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.CLI.dll' could not be found
27>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Parser.dll' could not be found
38>------ Build started: Project: Common.Tests.CSharp, Configuration: Debug x64 ------
30>'"F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/Common.Gen.exe"' is not recognized as an internal or external command,
30>operable program or batch file.
30>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(128,5): error MSB3073: The command ""F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/Common.Gen.exe"
30>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(128,5): error MSB3073: :VCEnd" exited with code 9009.
30>Done building project "Common.CLI.vcxproj" -- FAILED.
39>------ Build started: Project: StandardLib.CLI, Configuration: Debug x64 ------
40>------ Build started: Project: Common.Tests.CLI, Configuration: Debug x64 ------
36>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
36>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CSharp.CSharp.dll' could not be found
34>  '"F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/NamespacesDerived.Gen.exe"' is not recognized as an internal or external command,
34>  operable program or batch file.
34>C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(1330,5): error MSB3073: The command ""F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/NamespacesDerived.Gen.exe"" exited with code 9009.
41>------ Build started: Project: NamespacesDerived.CSharp, Configuration: Debug x64 ------
35>  '"F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/VTables.Gen.exe"' is not recognized as an internal or external command,
35>  operable program or batch file.
35>C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(1330,5): error MSB3073: The command ""F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/VTables.Gen.exe"" exited with code 9009.
42>------ Build started: Project: VTables.Tests.CSharp, Configuration: Debug x64 ------
31>'"F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/CLI.Gen.exe"' is not recognized as an internal or external command,
31>operable program or batch file.
31>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(128,5): error MSB3073: The command ""F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/CLI.Gen.exe"
31>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(128,5): error MSB3073: :VCEnd" exited with code 9009.
31>Done building project "CLI.CLI.vcxproj" -- FAILED.
43>------ Build started: Project: CLI.Tests.CLI, Configuration: Debug x64 ------
38>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\Common.CSharp.dll' could not be found
38>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
39>'"F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/StandardLib.Gen.exe"' is not recognized as an internal or external command,
39>operable program or batch file.
39>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(128,5): error MSB3073: The command ""F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/StandardLib.Gen.exe"
39>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(128,5): error MSB3073: :VCEnd" exited with code 9009.
39>Done building project "StandardLib.CLI.vcxproj" -- FAILED.
44>------ Build started: Project: StandardLib.Tests.CLI, Configuration: Debug x64 ------
42>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
42>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\VTables.CSharp.dll' could not be found
41>  '"F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/NamespacesDerived.Gen.exe"' is not recognized as an internal or external command,
41>  operable program or batch file.
41>C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(1330,5): error MSB3073: The command ""F:\temp\CppSharp\build\vs2017\lib\Debug_x64\/NamespacesDerived.Gen.exe"" exited with code 9009.
45>------ Build started: Project: NamespacesDerived.Tests.CSharp, Configuration: Debug x64 ------
37>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
37>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\Encodings.CSharp.dll' could not be found
44>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
44>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\StandardLib.CLI.dll' could not be found
40>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\Common.CLI.dll' could not be found
40>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
45>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
45>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\NamespacesBase.CSharp.dll' could not be found
45>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\NamespacesDerived.CSharp.dll' could not be found
43>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CLI.CLI.dll' could not be found
43>CSC : error CS0006: Metadata file 'F:\temp\CppSharp\build\vs2017\lib\Debug_x64\CppSharp.Generator.Tests.dll' could not be found
========== Build: 13 succeeded, 32 failed, 0 up-to-date, 0 skipped ==========
```

--- a/src/Generator/Types/Std/Stdlib.cs
+++ b/src/Generator/Types/Std/Stdlib.cs
@@ -155,16 +155,52 @@ namespace CppSharp.Types.Std
                 return new CustomType(typePrinter.IntPtrType);
             }
 
-            if (Context.Options.Encoding == Encoding.ASCII ||
-                Context.Options.Encoding == Encoding.UTF8)
+            Type type = Type.Desugar();
+
+            uint charWidth = 0;
+            if (type is PointerType pointerType)
+                charWidth = GetCharPtrWidth(pointerType);
+            else if (type.GetPointee()?.Desugar() is PointerType pointeePointerType)
+                charWidth = GetCharPtrWidth(pointeePointerType);
+
+            if (charWidth == 8)
+            { 
+                if (Context.Options.Encoding == Encoding.ASCII)
+                    return new CustomType("[MarshalAs(UnmanagedType.LPStr)] string");
                 return new CustomType("[MarshalAs(UnmanagedType.LPUTF8Str)] string");
+            }
 
             if (Context.Options.Encoding == Encoding.Unicode ||
                 Context.Options.Encoding == Encoding.BigEndianUnicode)
+            {
+                // NOTE: This will break if charWidth is not 16 bit which may
+                // happen if someone uses wchar_t on platforms where it's size is 32 bit.
+                // TODO: Create a custom marshaller to support that scenario.
+                // Once that's done consider supporting char32_t as well
                 return new CustomType("[MarshalAs(UnmanagedType.LPWStr)] string");
+            }
+
+            if (Context.Options.Encoding == Encoding.UTF8 && charWidth == 16)
+                return new CustomType("[MarshalAs(UnmanagedType.LPWStr)] string"); // fallback;
 
             throw new System.NotSupportedException(
                 $"{Context.Options.Encoding.EncodingName} is not supported yet.");
+        }
+
+        public uint GetCharPtrWidth(PointerType pointer)
+        {
+            var pointee = pointer?.Pointee?.Desugar();
+            if (pointee != null)
+            {
+                if (pointee.IsPrimitiveType(PrimitiveType.Char))
+                    return Context.TargetInfo.CharWidth;
+                if (pointee.IsPrimitiveType(PrimitiveType.WideChar))
+                    return Context.TargetInfo.WCharWidth;
+                if (pointee.IsPrimitiveType(PrimitiveType.Char16))
+                    return Context.TargetInfo.Char16Width;
+            }
+
+            return 0;
         }
 
         public override void CSharpMarshalToNative(CSharpMarshalContext ctx)

--- a/tests/CSharp/CSharp.Tests.cs
+++ b/tests/CSharp/CSharp.Tests.cs
@@ -1479,4 +1479,11 @@ public unsafe class CSharpTests : GeneratorTestFixture
     {
         StringAssert.EndsWith(nameof(CSharp.TestAnonymousMemberNameCollision._0.__0), "__0");
     }
+
+    [Test]
+    public void TestStringMarshall()
+    {
+        Assert.IsTrue(StringMarshall.CSharpString8(StringMarshall.CSharpString));
+        Assert.IsTrue(StringMarshall.CSharpString16(StringMarshall.CSharpString));
+    }
 }

--- a/tests/CSharp/CSharp.cpp
+++ b/tests/CSharp/CSharp.cpp
@@ -1777,3 +1777,13 @@ boolean_t takeTypemapTypedefParam(boolean_t b)
 {
     return b;
 }
+
+bool StringMarshall::CSharpString8(const char* in)
+{
+    return in[0] == 'C' && in[1] == '#';
+}
+
+bool StringMarshall::CSharpString16(const char16_t* in)
+{
+    return in[0] == 'C' && in[1] == '#';
+}

--- a/tests/CSharp/CSharp.h
+++ b/tests/CSharp/CSharp.h
@@ -1481,3 +1481,11 @@ struct TestVariableWithoutType
 
     static constexpr auto variable = create(n...);
 };
+
+class DLL_API StringMarshall
+{
+public:
+    static constexpr const char* CSharpString = "C#";
+    static bool CSharpString8(const char* in);
+    static bool CSharpString16(const char16_t* in);
+};

